semanticdb3/semanticdb3.md

```diff
@@ -0,0 +1,491 @@
+and we may decide to expand the format to accommodate namespacing rules not present in Scala.
+
+**Global symbols**. Correspond to a definition that can be referenced outside
+the compilation unit where the definition is defined.
+Global symbol format is a concatenation of signatures of the owner chain
+of the corresponding global definition, where:
+  * The owner chain of a definition is a list of its enclosing definitions
+    starting with the outermost one, with the outermost definition being
+    either `_empty_` (the special empty package [\[17\]][17]) or
+    `_root_` (the special root package [\[18\]][18]). For example,
+    for the standard `Int` class, the owner chain is `[_root_, scala, Int]`.
+  * The signature of a definition is:
+    * For a method, concatenation of its name, its JVM method descriptor [\[19\]][19]
+      and a dot (`.`). The JVM method descriptor is used to distinguish
+      overloaded methods as mandated by the Scala Language Specification [20].
+      For any other term definition (package, object, val or var),
+      concatenation of its name and a dot (`.`).
+    * For a type definition (class, type alias or type member),
+      concatentation of its name and a pound sign (`#`).
+
+For example, the standard `Int` class must be modelled by a global symbol `_root_.scala.Int#`.
+
+**Local symbols**. Correspond to a definition that isn't global (see above).
+Local symbol format is deliberately unspecified except for two restrictions:
+  * Local symbols must start with `local`, so that they can be easily
+    distinguished from global symbols.
+  * Local symbols must be unique within the underlying document.
+
+For example, `x` in `def identity[T](x: T): T` may be modelled by local symbols
+`local0`, `local_x`, `local_identity_x`, as long as these names are unique within
+the underlying document. The same logic applies to the type parameter `T`,
+which is also a local definition.
+
+**Placeholder symbols**. Are used to model original snippets of code in [Synthetics](#synthetic).
+Must not be used outside `Synthetic.text` documents. Placeholder symbols are
+always equal to an asterisk (`*`).
+
+### Identifier
+
+```protobuf
+message Identifier {
+  Range range = 1;
+  string symbol = 2;
+  int64 role = 3;
+}
+```
+
+"Identifiers" is a section of a [TextDocument](#textdocument) that represents
+the results of name resolution for identifiers in the underlying snippet of code.
+
+`Identifier` refers to a [Range](#range) in the code and has a symbol as explained
+in [Symbol](#symbol). `role` is an enumeration that describes the semantic role
+that the identifier performs in the snippet of code. Like many other enumerations
+in SemanticDB, this one is usecase-driven and will likely be updated in the future.
+
+<table>
+  <tr>
+    <td><b>Value</b></td>
+    <td><b>Explanation</b></td>
+  </tr>
+  <tr>
+    <td><code>1 << 1</code></td>
+    <td>Reference</td>
```

Comment:
How is `1 << 1` Related to reference?

Reply:
ahhh, It's a bit shift. I would just put 1, 2, 4, ...

Reply:
The idea is that a symbol occurrence can have multiple roles. Currently, we only have reference/definition but other examples that come to mind include `extends`. See https://youtu.be/jGJhnIT-D2M?t=10m12s how they use this role information in xcode https://youtu.be/jGJhnIT-D2M?t=10m12s

Reply:
Another potential usecase for role would be to label 

```scala
package scala {
  val Left = scala.util.Left
}
```
as an `alias` so that it's easy to "Find references" for `scala.util.Left` and include results for `scala.Left`

Reply:
I like the alias idea more and more every day.